domain/src/test/java/org/oppia/domain/profile/ProfileManagementControllerTest.kt

```diff
@@ -95,8 +122,392 @@ class ProfileManagementControllerTest {
+    profileManagementController.setCurrentProfileId(profileId).observeForever(mockUpdateResultObserver)
+
+    verify(mockUpdateResultObserver, atLeastOnce()).onChanged(updateResultCaptor.capture())
+    assertThat(updateResultCaptor.value.isFailure()).isTrue()
+  }
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testGetProfileId_addProfiles_setProfile_checkGetProfileIdIsCorrect() = runBlockingTest(coroutineContext) {
+    addTestProfiles()
+    advanceUntilIdle()
+
+    val profileId = ProfileId.newBuilder().setInternalId(2).build()
+    profileManagementController.setCurrentProfileId(profileId).observeForever(mockUpdateResultObserver)
+    advanceUntilIdle()
+
+    verify(mockUpdateResultObserver, atLeastOnce()).onChanged(updateResultCaptor.capture())
+    assertThat(updateResultCaptor.value.isSuccess()).isTrue()
+    assertThat(profileManagementController.getCurrentProfileId().internalId).isEqualTo(2)
+  }
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testLoginToProfile_addProfiles_loginToProfile_checkGetProfileIdAndLoginTimestampIsCorrect() = runBlockingTest(coroutineContext) {
+    addTestProfiles()
+    advanceUntilIdle()
+
+    val profileId = ProfileId.newBuilder().setInternalId(2).build()
+    profileManagementController.loginToProfile(profileId).observeForever(mockUpdateResultObserver)
+    advanceUntilIdle()
+    profileManagementController.getProfile(profileId).observeForever(mockProfileObserver)
+
+    verify(mockUpdateResultObserver, atLeastOnce()).onChanged(updateResultCaptor.capture())
+    verify(mockProfileObserver, atLeastOnce()).onChanged(profileResultCaptor.capture())
+    assertThat(updateResultCaptor.value.isSuccess()).isTrue()
+    assertThat(profileResultCaptor.value.isSuccess()).isTrue()
+    assertThat(profileManagementController.getCurrentProfileId().internalId).isEqualTo(2)
+    assertThat(profileResultCaptor.value.getOrThrow().lastLoggedInTimestampMs).isNotEqualTo(0)
+  }
+
+  private fun addTestProfiles() {
+    profilesList.forEach {
+      profileManagementController.addProfile(it.name, it.pin, null, it.allowDownloadAccess)
+    }
+  }
+
+  private fun checkTestProfilesArePresent(resultList: List<Profile>) {
+    profilesList.forEachIndexed { idx, profile ->
+      assertThat(resultList[idx].name).isEqualTo(profile.name)
+      assertThat(resultList[idx].pin).isEqualTo(profile.pin)
+      assertThat(resultList[idx].allowDownloadAccess).isEqualTo(profile.allowDownloadAccess)
+      assertThat(resultList[idx].id.internalId).isEqualTo(idx)
+      assertThat(File(getAbsoluteDirPath(idx.toString())).isDirectory).isTrue()
+    }
+  }
+
+  private fun getAbsoluteDirPath(path: String): String {
+    /**
+     * context.filesDir.toString() looks like /tmp/robolectric-Method_test_name/org.oppia.util.test-dataDir/files
+     * dropLast(5) removes files from the path and then it appends the real path with "app_" as a prefix
+     */
+    return context.filesDir.toString().dropLast(5) + "app_" + path
```

Comment:
Do we actually need to do this? I'd expect the prod and test code to be agnostic of the path format.

Reply:
I think this is needed because getDir in Robolectric creates directories in /tmp/robolectric-Method_test_name/org.oppia.util.test-dataDir/app_dirname. I needed to remove the "files" part of context.filesDir.toString() and add the prefix app_ to get the absolute directory.